### ISSUE TITLE
Support for PostgreSQL as a broker - gauge of interest

### ISSUE THREAD

**Issue by jasiek** [celery/celery#5149]
PostgreSQL has a LISTEN/NOTIFY mechanism which appears to be capable of implementing a message queue-like functionality (likely not all by itself, as it may require some supporting infrastructure in the form of a table of messages). 

Is there any interest in using PostgreSQL as both a broker and a result store (which I believe is currently fully supported)? It looks attractive to me as it would reduce the number of different components in my system (ie why run redis/rabbitmq on the side, when I can store everything in my database).

CC: @xirdneh

**Comment by auvipy** [comment]
already we use django and sqlalchemy as broker and result backend for the development environment. but a relational database is not a good option for production level broker, you can certainly do it at your own risk and create a separate package for that to demonstrate your experiment. the main issue is scaling, you can use whatever you want in your production system, and before qualifying anything to be part of celery core that should have proven ability and high interest in the community.

**Comment by jasiek** [comment]
> celery core that should have proven ability and high interest in the community.

My point exactly - this is to gauge interest. May we leave this ticket open for some time to see if there's any?

**Comment by auvipy** [comment]
celery is a 10 years old project. if using a database as a broker was a good idea, the request would be more frequent. if you have a dire need, please create your own package.

**Comment by thedrow** [comment]
I agree. We had that option, it wasn't viable and we don't have the spare time to support it.

**Comment by bersace** [comment]
@jasiek for your information, I started a Postgres broker for dramatiq : https://gitlab.com/dalibo/dramatiq-pg . It's early, but I think that using Postgres as a broker could be an option for small configuration. Say you want to run a small webapp on RaspberryPi.

**Comment by jasiek** [comment]
Thanks @bersace - this is exactly what I had in mind.

**Comment by bersace** [comment]
@jasiek you're welcome.

Note also that Skype implemented a high performance queue broker in Postgres as an extension named [pgq](https://github.com/pgq/pgq) . It's there for ages. I'm willing to investigate with this as an alternative to rabbitmq.

**Comment by chukkwagon** [comment]
How did the celery team establish that database backed job queues were a bad idea? Is that "common knowledge" or something that has been tested and demonstrated conclusively? There are some benefits to database backed job queues, particularly in the way of co-transactionality guarantees. It's also arguably easier to reason about architecturally, depending on how confident you are in your ability to maintain and scale multiple independent data stores.

>celery is a 10 years old project. if using a database as a broker was a good idea, the request would be more frequent

Whew, lots to unpack here! Databases have changed and improved quite a bit in the last 10 years, even old faithfuls like PostgreSQL. There's nothing wrong with revisiting first principles every once in a while :)

This is something I've hoped celery would support since I first started using it 4 years ago, so make that two fools in favor of this hopelessly un-scalable option. I'm glad you brought it up @jasiek ! If you're interested in learning a little more about an engineer might consider a database backed queue, here's a link to a talk by the VP of Architecture at Betterment: https://www.infoq.com/presentations/betterment-delayed-job on their (RoR) implementation

**Comment by auvipy** [comment]
You are free to have your own 3rd party package for your need. No everything needs to be in the core. We have tons of backlogs to do in the next 2-3 years already.

**Comment by aljungberg** [comment]
I don't have a horse in this race, found this conversation by chance. Just to put it out there, there are two features that make Postgres a much better job broker than conventional wisdom suggests.

- 'SKIP LOCKED` to allow multiple workers to select from a job list with low locking overhead [1](https://www.postgresql.org/docs/13/sql-select.html)
- `NOTIFY` pubsub functionality to enable workers to idle with low CPU draw while still waking up immediately for new work [2](https://www.postgresql.org/docs/13/sql-notify.html)

I suspect it's possible to build a high performant distributed task queue based on Postgres, taking advantage of transactional safety for low overhead 'execute exactly once' robustness.

**Comment by auvipy** [comment]
> I suspect it's possible to build a high performant distributed task queue based on Postgres, taking advantage of transactional safety for low overhead 'execute exactly once' robustness.

yeah that could be a possibility

**Comment by auvipy** [comment]
@xirdneh I am reopening it to have more input from community

**Comment by odigity** [comment]
I'm interested in this.  It would be nice if Tasks were a first-class participant in my application's data model, so I could easily query task status and results alongside the business objects related to the task, perhaps to display in an Admin UI.  Also, one less database to setup and operate.

Are there any prototypes of a Celery-Broker adapter for Postgres in the wild to try out or take inspiration from?

**Comment by Idimmusix** [comment]
I support the issue of using postgresql as a broker, at least for small projects.

**Comment by palmy-investing** [comment]
> I support the issue of using postgresql as a broker, at least for small projects.

When would you deny using it as broker. Please define small project, I am not sure whether to use redis/rabbitmq or postgres. Thanks

**Comment by ddahan** [comment]
I'm not sure to understand how it is currently supported:
- On one hand: nothing in the documentation, and this issue is open
- On the other hand: this [Stack Overflow answer](https://stackoverflow.com/a/47604045/2255491) thats shows a working method to do it, and sounds pretty clean.

**Comment by odoublewen** [comment]
@jasiek Could you please elaborate on your comment?  Because, to my admittedly naive self, "retain information about jobs and implement queues" sounds like what I need for a low-throughput task queue.  

When I set up a sandbox instance of Celery 5.2.7 using:

```
    broker_url = 'sqla+postgresql+psycopg2://127.0.0.1/celery'
    result_backend = 'db+postgresql://127.0.0.1/celery'
```

...as far as I can tell, everything works fine.  I can create tasks, my worker picks them up and processes them, and I can see the results when they're done.

Are there important caveats/warnings about using Celery like this?  Should it be "safe enough", if "low-throughput" means, say, ~1 job per hour?

**Comment by malthe** [comment]
I would put low-throughput at 1 per second or less in this context. That's fine for PostgreSQL to handle.

**Comment by aljungberg** [comment]
I mean Postgres is extremely fast. With good tombstone management you can probably write thousands of tasks per second on a medium machine, no sweat. 

I think the `LISTEN/NOTIFY` angle is critical though because otherwise the worker task selection might become a CPU drain. It did for us anyhow (not with Celery but an internal producer/consumer system). Never really dug into why but I suspect as efficient as `SELECT ... SKIP LOCKED` can be there must still be some time complexity linear to the number of pages visited to look for unlocked rows. (Possibly we just had a lot of dead tuples in our case because after a big "delete finished tasks" and a vacuum the problem went away for us.)

Anyhow point is you probably don't want all the Celery workers to be constantly hammering the queue table for new work because those selects, even if fast, still have a cost related to the data composition of the table. You might want a combination of push and poll in fact.

1. Start a new worker by looking for work. Something like,`SELECT task_id FROM tasks WHERE state='pending' AND worker IS NULL ORDER BY task_id FOR UPDATE SKIP LOCKED LIMIT 1`. If a task is found, go to 4.
2. If no work is available, `LISTEN` and sleep.
3. If a notification comes in for task `123`, wake up. Opportunistically assume `123` will be this worker's next task. Stop listening.
5. Try to take ownership of the task with an `UPDATE tasks SET state='running', worker=myworkerid WHERE task_id=123 AND worker IS NULL` (pseudocode). If we came from 1, this will always succeed (because we got a lock in our select). If we came from 3, there may be dogpiling here if many workers are idle. Only 1 will win thanks to ACID safety and thanks to filtering on the primary key every attempt will execute with minimal cost. If the result is 0 updates, this worker didn't win. Go to 1.
6. Perform the task, then `DELETE FROM tasks WHERE task_id=123; INSERT INTO task_tombstones ...`
7. Go to 1.

Notes on FIFO (first in, first out):

- Normally one expects tasks to be executed in the order they were scheduled.
- This algorithm is "approximately FIFO".
- In the`SELECT`, the `ORDER BY task_id` guarantees FIFO behaviour (assuming always increasing task ids).
- The approximate part is because we mix in `LISTEN/NOTIFY`. 
- The `LISTEN` mechanism *roughly* preserves FIFO: with the above strategy you only ever get idle workers when the queue is empty. If a new task comes in and we NOTIFY and workers wake up to handle it, well there's only one task in the queue so it's FIFO in any order.
- ...however when tasks are created you'd issue the `NOTIFY` after the `COMMIT` to guarantee the row is visible to the workers. So there's a race to be the first to notify when multiple tasks are created around the same time such that the order of tasks might end up different from the order of notifications.
- The ordering is still approximately right since the time between `NOTIFY` and `COMMIT` will be small. We're talking about a bunch of tasks created like the same millisecond.
- ...but also a `NOTIFY` might be sent in the time window after a worker did its select but before it started to `LISTEN`. Then another `NOTIFY` for a subsequent task wakes it up and it ends up working on the second task in the queue, not the first. When it's done it will go to step 1 and end up picking the first task.
- ...so in this case the tasks end up being performed in reverse order and there could be a delay between when the first task is missed and the next `NOTIFY` comes in. See "lost notification edge case" below for how we reduce this delay.

Other notes:

- In step 6 we move the completed task to a separate table (delete and insert).
- It's an optimisation to keep a small active tasks table. I haven't run any numbers on this, but one suspects that even with proper indexing on `state` (and maybe `worker`), `SELECT ... FOR UPDATE SKIP LOCKED` will run faster with a smaller set of active pages to work through.
- There's no data safety risk here (where a task can be completed but not get a tombstone) due to ACID. You run the delete and insert in the same transaction so one can't happen without the other.
- When a worker has processed a `NOTIFY`, whether it did some work or not, it does a full `SELECT` after to look for more work before returning to its idle state. Because we stop listening after receiving a message, going back to idling immediately after work would be incorrect. More work might have come in while we weren't listening and now we'd have an idle worker despite enqueued work.

Lost notification edge-case:

There needs to be a time-out for idle workers (with some random jitter) after which they go back to looking for work through a regular `SELECT`. There is a small window where we go from no work to having work right in that space between step 1 and 2, before any worker starts to 'LISTEN'. 

It should be hard to trigger that case if you have more than 1 worker since it'd be unlikely all workers would be between step 1 and 2 at once. Even if you do trigger the case, you might be "rescued" by another task coming in soon thereafter which would wake everyone up.

But let's say you do manage to hit just the right timing and no new tasks are coming at the moment. A time-out sets an upper limit for how long this could last, at the cost of adding some unnecessary `SELECTs`.

Alternatively, one could go without a timeout and instead make handling this edge case part of the general Celery heartbeat maintenance stuff because there are other edge cases which must be handled too. Like when a worker freezes/crashes and never finishes a task and we need to mark it as failed or set it up to retry with another worker. So when we're in there looking for hung tasks we might as well finish by NOTIFYING on the first good, unstarted task we saw while checking all tasks.

1. If all workers are busy, the notification will just be ignored at nearly no cost.
2. But if we're in this rare edge case with a "lost task" and all sleeping workers, it will be resolved.
3. And there will be no unnecessary `SELECT`s since we have to do this maintenance anyhow and if we don't find any pending task we don't notify, so all the workers just keep on sleeping.

Either way we get an upper limit on how long this edge case can delay a task's execution.

(You could also avoid the edge case entirely by doing a final check, a select, immediately after starting to `LISTEN`. This eliminates the window when notifications can be lost. But I haven't thought about this hard enough to be absolutely sure there aren't other edge cases, so I'd lean towards the time-out/heartbeat solution since it'll provide a failsafe.)

**Comment by auvipy** [comment]
I'm now positive about the idea but we need to figure out best implementation and design detail. this will surely reduce a lots of complexity and ease the entry to celery. I might try a proof of concept implementation in near future.

**Comment by auvipy** [comment]
One interesting thing is we have some degree of experimental support through SQLA already here https://github.com/celery/kombu/tree/main/kombu/transport/sqlalchemy but I think that needs a lot of improvement. and we can iterate over it or come up with a better approach then that one.

**Comment by malthe** [comment]
@auvipy problem is, it's the database-specific features (and here I am thinking mostly of postgres) that make this possible – can't really use a generic backend and expect the same outcome.

**Comment by auvipy** [comment]
> @auvipy problem is, it's the database-specific features (and here I am thinking mostly of postgres) that make this possible – can't really use a generic backend and expect the same outcome.

yes that is of course very true. I 100% agree with you. we can start with SQLA and then make pg specific more direct codes and drivers, I guess

**Comment by kuhella** [comment]
It's 2024 and this feature is still very important

**Comment by emer-bestseller** [comment]
Is there any progress on this concept? I sounds like a great idea to me!

**Comment by nileshtrivedi** [comment]
More prior art for SQL database as job queue: https://github.com/rails/solid_queue

**Comment by jacopofar** [comment]
We use in production [this task queue](https://github.com/flix-tech/postgres-tq) we implemented, based on Postgres. The [upcoming version](https://github.com/flix-tech/postgres-tq/pull/10) will probably add the possibility to postpone tasks. It should be feasible to add a wrapper to use it as a broker.

**Comment by nileshtrivedi** [comment]
Relevant project: https://github.com/janbjorge/PgQueuer

**Comment by fangpenlin** [comment]
I am also thinking about using a feature like this, mainly because I want the task to be "queued" with my transaction when I commit all other entries into different tables. I found there are also other task worker queue implementation using PostgreSQL's skip lock such as this one

https://github.com/hatchet-dev/hatchet

**Comment by fangpenlin** [comment]
We just built a new framework called BeanQueue for our internal use. It's based on SKIP LOCKED and LISTEN / NOTIFY statements. We use SQLAlchemy as the ORM. You can check out the project here:

https://github.com/LaunchPlatform/bq

**Comment by matthew-mcallister** [comment]
Whatever solution is implemented, I think the most important feature is that tasks are queued in the same transaction as the calling code. The DB connection should be shared between Celery and the application. Without transaction support, there is far less motivation to use PostgreSQL over to Redis/RabbitMQ.

**Comment by nileshtrivedi** [comment]
FYI, [DEP 0014](https://github.com/django/deps/blob/main/accepted/0014-background-workers.rst) (Django Enhancement Proposal) is proposing support for background workers within Django:

**Comment by slavakurilyak** [comment]
Here's a great writeup on choosing [Postgres queue technology](https://adriano.fyi/posts/2023-09-24-choose-postgres-queue-technology/) which will in making a decision

My favourite insight is at the end:

> Make boring technology your default choice, and choose Postgres if it bores you

**Comment by ddelange** [comment]
fwiw these requirements work
```
psycopg[binary,pool]~=3.2.3
celery[sqlalchemy]~=5.4.0  # installs sqlalchemy v2 which supports psycopg v3
```
with a broker url: `sqlalchemy+postgresql+psycopg://<db_uri>/celery`

ref kombu: https://docs.celeryq.dev/projects/kombu/en/v5.4.0/reference/kombu.transport.sqlalchemy.html
ref sqlalchemy: https://docs.sqlalchemy.org/en/20/dialects/postgresql.html

**Comment by ocervell** [comment]
Can anyone confirm this is working well in production setups with hundreds of transactions ?
Planning to replace Redis backend + RabbitMQ broker with just Postgres since it's the goat.

**Comment by ddelange** [comment]
we were pretty happy with not an insane amount of throughput (order of 100k task executions per day). you might have to tweak vacuum / WAL configs to keep postgres happy.

**Comment by ebertti** [comment]
For those landing here — I published [`kombu-pgmq`](https://github.com/ebertti/kombu-pgmq), a Kombu transport that uses [PGMQ](https://github.com/pgmq/pgmq) (the PostgreSQL message queue extension) as the Celery broker. Already on PyPI, no Redis or RabbitMQ needed.

```bash
pip install kombu-pgmq[pgmq]
```

```python
app = Celery(
    "myapp",
    broker_transport="kombu_pgmq.transport:TransportPGMQ",
    broker_url="postgresql://user:pass@localhost:5432/mydb",
)
```

Uses PGMQ's visibility timeout for redelivery (same semantics as SQS, so reject/retry works out of the box). Two backends: `TransportPGMQ` (official `pgmq` client) and `TransportPsycopg` (raw psycopg 3, smaller footprint). Full test suite runs against a real Postgres+PGMQ instance.

Re the transaction concern raised upthread: PGMQ's `pgmq.send()` runs inside your Postgres transaction, so you can enqueue a task atomically alongside your business logic rows — that's a big part of PGMQ's value prop over Redis/RabbitMQ.

**Comment by auvipy** [comment]
we got a PR supporting PGMQ to kombu and celery which might land in v5.7.

**Comment by tavallaie** [comment]
To fully support of pgmq in celery. Do I need to make PR to this repo too? 
I am still new hwre and trying to find my way.

**Comment by auvipy** [comment]
may be no... but for using pgmq as results back end might be some change needed in celery... you can first check the sqlalchemy database backend code here. also we should reference the merged pr in kombu here

**Comment by tavallaie** [comment]
this the PR already merged in Kombu: 
https://github.com/celery/kombu/pull/2559
I will take look on how we can use it as result backend in celery too.

**Comment by auvipy** [comment]
beside the database backend checking the base back end related code might be also needed https://github.com/celery/celery/blob/main/celery/backends/base.py

**Comment by tavallaie** [comment]
I checked both implementations. I don’t think using PGMQ as the result backend is a good idea. We could instead implement a backend purely with psycopg so we don’t depend on SQLAlchemy. Do you think that’s worth implementing?

**Comment by tavallaie** [comment]
Also I can add Peek function to pgmq and this way we can use it as result backend too. 
also it is possible to install pgmq extension on postgres directly from python library we already have documentation for that.

**Comment by auvipy** [comment]
cant we just use SQLALChemy initially? with psycopg initially? and see how it goes? I think that will let use postgresql as both broker and result back end bypassing SQLA atleast as broker

**Comment by tavallaie** [comment]
Fine. SQLAlchemy + psycopg for results, PGMQ for the broker. I’ll add docs/tests in celery for that pairing if you want.

**Comment by auvipy** [comment]
yes, just adding a doc will help us close this

**Comment by auvipy** [comment]
> Fine. SQLAlchemy + psycopg for results, PGMQ for the broker. I’ll add docs/tests in celery for that pairing if you want.

with your docs pr, we can close this issue

**Comment by tavallaie** [comment]
sorry I was busy with work, I will make a PR this weekend.
